api/src/main/java/org/apache/iceberg/PartitionSpec.java

```diff
@@ -331,7 +331,11 @@ private Builder(Schema schema) {
       this.schema = schema;
     }
 
-    private void checkAndAddPartitionName(String name) {
+    private void checkAndAddPartitionName(String name, String transform) {
+      if (!transform.equalsIgnoreCase("identity")) {
```

Comment:
I think the test case for this points out anther case that shouldn't be allowed. Creating an identity partition named after the wrong field.

What about updating this by passing in `Integer identitySourceColumnId`. Then the validation should be if there is a schema with the field name and the ID doesn't match `identitySourceColumnId` then the exception is thrown. Like this:

```java
  NestedField schemaField = schema.findField(name);
  Preconditions.checkArgument(schemaField == null || schemaField.fieldId() == identitySourceColumnId,
      "Cannot create partition with the same name as schema field: %s", name);
```

Reply:
Done. Only change I made to your logic is I handled both cases in separate if condition blocks as the identity case is unique from the other transform cases.